rewriter/ChainedSig.cc

```diff
@@ -0,0 +1,184 @@
+        auto blockBody = ast::cast_tree<ast::Send>(block->body);
+
+        // If the blockBody is not a send, then we have a sequence of expressions inside the signature block
+        if (blockBody == nullptr) {
+            if (auto e = ctx.beginError(block->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                e.setHeader("Malformed signature: cannot have multiple statements inside a signature block");
+            }
+
+            return tree;
+        }
+
+        return buildReplacement(ctx, send, blockBody);
+    }
+
+    bool firstSendIsSig(ast::Send *send) {
+        do {
+            if (send->fun == core::Names::sig()) {
+                this->previousSigSend = send;
+                return true;
+            }
+            send = ast::cast_tree<ast::Send>(send->recv);
+        } while (send);
+
+        return false;
+    }
+
+    void checkDuplicates(core::MutableContext ctx, ast::Send *sendFromSigBlock, core::NameRef sendCopyFun,
+                         core::LocOffsets sendCopyFunLoc) {
+        while (sendFromSigBlock) {
+            if (sendFromSigBlock->fun == sendCopyFun) {
+                if (auto e = ctx.beginError(sendFromSigBlock->loc, core::errors::Rewriter::InvalidChainedSig)) {
+                    e.setHeader("Duplicate invocation of `{}` in signature declaration", sendCopyFun.toString(ctx));
+                    e.addErrorLine(ctx.locAt(sendCopyFunLoc), "Initial invocation made here");
+                }
+            }
+
+            sendFromSigBlock = ast::cast_tree<ast::Send>(sendFromSigBlock->recv);
+        }
+    }
+
+    bool invalidChainedStatement(core::MutableContext ctx, ast::Send *send) {
+        if (send->fun == core::Names::params() || send->fun == core::Names::returns() ||
+            send->fun == core::Names::void_() || send->fun == core::Names::bind() ||
+            send->fun == core::Names::checked() || send->fun == core::Names::onFailure() ||
+            send->fun == core::Names::typeParameters()) {
+            if (auto e = ctx.beginError(send->funLoc, core::errors::Rewriter::InvalidChainedSig)) {
+                e.setHeader("Cannot use `{}` outside of a sig block", send->fun.toString(ctx));
+            }
+
+            return true;
+        }
+
+        return false;
+    }
+
+    ast::ExpressionPtr buildReplacement(core::MutableContext ctx, ast::Send *send, ast::Send *blockBody) {
+        ast::Send::ARGS_store args;
+        auto treeCopy = send->deepCopy();
+        auto sendCopy = ast::cast_tree<ast::Send>(treeCopy);
+
+        // Go through each part of the chained sig and make sure it's not duplicated inside the block
+        // E.g.: `sig.abstract { abstract.void }`
+        while (sendCopy && sendCopy->fun != core::Names::sig()) {
+            checkDuplicates(ctx, blockBody, sendCopy->fun, sendCopy->funLoc);
+
+            sendCopy = ast::cast_tree<ast::Send>(sendCopy->recv);
+        }
+
+        sendCopy = ast::cast_tree<ast::Send>(treeCopy);
```

Comment:
Can we use a separate, named variable for the loop iterator here so that we don't have to re-assign `sendCopy` after the loop, please?